src/pickBy.ts

```diff
@@ -1,53 +1,186 @@
-import { keys } from "./keys";
+import { type IfNever, type Simplify } from "type-fest";
+import {
+  type EnumerableStringKeyOf,
+  type EnumerableStringKeyedValueOf,
+} from "./_types";
 import { purry } from "./purry";
 
+// Because pickBy needs to iterate over all entries of the object, only
+// enumerable keys (those returned by `Object.entries`) could be part of the
+// output object. Any symbol keys would be skipped and filtered out.
+type EnumerableKey<T> = `${T extends number | string ? T : never}`;
+
+// When the predicate isn't a type-guard we don't know which properties would be
+// part of the output and which wouldn't so we can only safely downgrade the
+// whole object to a Partial of the input.
+type EnumeratedPartial<T> = Simplify<{
+  -readonly [P in keyof T as EnumerableKey<P>]?: Required<T>[P];
+}>;
+
+// When the predicate is a type-guard we have more information to work with when
+// constructing the type of the output object. We can safely remove any property
+// which value would never come up true for the predicate, AND we can also
+// assume that properties that match the predicate perfectly would **always**
+// show up in the output object. Hence to build the output object we need to
+// build and merge 2 output objects: One for the properties which have a value
+// of at type that would always yield a `true` result from the predicate, these
+// are the "matches", which would not change the "optionality" of the input
+// object's props, and one for partial matches which would also make the props
+// optional (as they could have a value that would be filtered out).
+type EnumeratedPartialNarrowed<T, S> = Simplify<
+  ExactProps<T, S> & PartialProps<T, S>
```

Comment:
**nit:** `AlwaysPickedProps` and `OptionalPickedProps` maybe?

then we can shorten the comment a bit...


```ts
// When the predicate is a type guard (`value is S`), each property will either:
// - never be S, which we always remove
// - always match S, which we always include (`AlwaysPickedProps`)
// - sometimes match S, which we make optional (`OptionalPickedProps`)
```

Reply:
(and something similar for `omitBy`, though i'm not sure what exactly)

Reply:
Sorry, I missed this when merging the PR, i'll take a look and maybe send a PR to fix